firehose-writer/src/main/java/com/expedia/www/haystack/pipes/firehoseWriter/S3Sender.java

```diff
@@ -0,0 +1,148 @@
+    private final FirehoseConfigurationProvider firehoseConfigurationProvider;
+    private final Factory factory;
+    private final FirehoseCountersAndTimer firehoseCountersAndTimer;
+    private final AmazonKinesisFirehoseAsync amazonKinesisFirehoseAsync;
+    private final Logger logger;
+    private final Counter throttledCounter;
+    private final FailedRecordExtractor failedRecordExtractor;
+
+    @Autowired
+    public S3Sender(FirehoseConfigurationProvider firehoseConfigurationProvider,
+                    Factory factory,
+                    FirehoseCountersAndTimer firehoseCountersAndTimer,
+                    AmazonKinesisFirehoseAsync amazonKinesisFirehoseAsync,
+                    Logger s3SenderLogger,
+                    Counter throttledCounter,
+                    FailedRecordExtractor failedRecordExtractor) {
+        this.firehoseConfigurationProvider = firehoseConfigurationProvider;
+        this.factory = factory;
+        this.firehoseCountersAndTimer = firehoseCountersAndTimer;
+        this.amazonKinesisFirehoseAsync = amazonKinesisFirehoseAsync;
+        this.logger = s3SenderLogger;
+        this.throttledCounter = throttledCounter;
+        this.failedRecordExtractor = failedRecordExtractor;
+    }
+
+    void sendRecordsToS3(final List<Record> records,
+                         final RetryCalculator retryCalculator,
+                         final Sleeper sleeper,
+                         final int retryCount,
+                         final Semaphore parallelism) {
+        final String streamName = firehoseConfigurationProvider.streamname();
+        final PutRecordBatchRequest request = factory.createPutRecordBatchRequest(streamName, records);
+        final Stopwatch stopwatch = firehoseCountersAndTimer.startTimer();
+        final int sleepMillis = retryCalculator.calculateSleepMillis();
+        boolean isInterrupted = false;
+
+        // If the main thread is interrupted while sleeping, the best behavior at this point in the code is to try
+        // to send the recordList to S3 (because not sending them could result in lost spans in Athena) but still stop
+        // execution; see https://www.ibm.com/developerworks/library/j-jtp05236/index.html. So we'll remember that an
+        // interrupt was requested and then reinterrupt after sending our request.
+        try {
+            factory.createSleeper().sleep(sleepMillis);
+        } catch (InterruptedException e) {
+            isInterrupted = true;
+        } finally {
+            amazonKinesisFirehoseAsync.putRecordBatchAsync(request,
+                    factory.createFirehoseAsyncHandler(this, stopwatch, request, sleepMillis, retryCount,
+                            records, retryCalculator, sleeper, parallelism, throttledCounter,
+                            failedRecordExtractor));
+        }
+        if(isInterrupted) {
+            factory.currentThread().interrupt();
```

Comment:
I don't think this ever gets executed. https://stackoverflow.com/questions/23919894/whats-the-difference-between-code-inside-finally-block-and-code-after-finally-b